File: software/ios/ble-rc/ConnectViewController.m
#import "ConnectViewController.h"
#import "DiscoveredRC.h"
#import "CarsController.h"
#import "SettingsViewController.h"
#import "CarViewController.h"
#import "StatusViewController.h"

@implementation ConnectViewController
{
    UIAlertView* searchDialog;
    UIAlertView* connectDialog;
    NSMutableArray* devices;
    bool searchEnabled;
}

-(id)initWithNibName:(NSString*)nibNameOrNil bundle:(NSBundle*)nibBundleOrNil
{
    self = [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];

    if (self)
    {
        self->devices = [[NSMutableArray alloc] init];
    }
    
    return self;
}

-(void)viewDidLoad
{
    [super viewDidLoad];
    
    [self searchClicked:self];
    
    [self.tabBarController setDelegate:self];
}

-(BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
    return (interfaceOrientation == UIInterfaceOrientationPortrait);
}

-(IBAction)searchClicked:(id)sender
{
    if (nil == self->searchDialog)
    {
        self->searchDialog = [[UIAlertView alloc]initWithTitle:@"Search" message:@"Searching for RC cars" delegate:nil cancelButtonTitle:@"Hide" otherButtonTitles:nil];

        [self->searchDialog show];
    
        /* if (nil != self->searchDialog)
        {
            UIActivityIndicatorView* indicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhiteLarge];
        
            indicator.center = CGPointMake(self->searchDialog.bounds.size.width / 2, self->searchDialog.bounds.size.height - 45);
            
            [indicator startAnimating];
            
            [self->searchDialog addSubview:indicator];
        } */
        
        //UIActivityIndicatorView* indicator = [[UIActivityIndicatorView alloc] initWithFrame:CGRectMake(0, 0, 20, 20)];
        //UIBarButtonItem* barButton = [[UIBarButtonItem alloc] initWithCustomView:indicator];
        //self.navigationItem.rightBarButtonItem = barButton;
        //[indicator startAnimating];
    }
    
    //if (nil == self->devices)
    {
        self->devices = [[NSMutableArray alloc] init];
    }
    
    [CarsController getInstance].delegateDiscover = self;
    [[CarsController getInstance] startDeviceSearch];
    
    [self->tableView reloadData];
}

-(IBAction)connectClicked:(id)sender
{
 //   [self deviceFound:[[DiscoveredRC alloc] initWithFakeData:@"beemmer"]];
 //   [self deviceFound:[[DiscoveredRC alloc] initWithFakeData:@"Italia"]];
 //   [self deviceFound:[[DiscoveredRC alloc] initWithFakeData:@"Drifter"]];
 //   [self deviceFound:[[DiscoveredRC alloc] initWithFakeData:@"Buggy"]];
 //   [self deviceFound:[[DiscoveredRC alloc] initWithFakeData:@"BLE-RC"]];
    
    NSIndexPath* selectedIndexPath = [self->tableView indexPathForSelectedRow];

    if (nil != selectedIndexPath)
    {
        if ((selectedIndexPath.row >= 0) && (selectedIndexPath.row < self->devices.count))
        {
            DiscoveredRC* device = [self->devices objectAtIndex:selectedIndexPath.row];
            
            if (nil != device)
            {
                if (nil == self->searchDialog)
                {
                    self->connectDialog = [[UIAlertView alloc]initWithTitle:@"Connect" message:@"Connecting to RC..." delegate:nil cancelButtonTitle:nil otherButtonTitles:nil];
                    
                    [self->connectDialog show];
                    
                    if (nil != self->connectDialog)
                    {
                        UIActivityIndicatorView* indicator = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhiteLarge];
                        
                        indicator.center = CGPointMake(self->connectDialog.bounds.size.width / 2, self->searchDialog.bounds.size.height - 45);
                        
                        [indicator startAnimating];
                        
                        [self->connectDialog addSubview:indicator];
                    }
                }
                
                [[CarsController getInstance] connect:device];
            }
        }
    }
    
    NSLog(@"Connect");
}

-(void)deviceFound:(DiscoveredRC*)device
{
    [self->devices addObject:device];
    [self->tableView reloadData];
}

-(void)searchEnded
{
    [self->tableView reloadData];
    [self->searchDialog dismissWithClickedButtonIndex:0 animated:true];
    self->searchDialog = nil;
    
    if (true == self->searchEnabled)
    {
        //[self searchClicked:self];
    }
}

-(void)connected
{
    [self->connectDialog dismissWithClickedButtonIndex:0 animated:true];
    self->connectDialog = nil;
    self.tabBarController.selectedIndex = 1;
    
    self->searchEnabled = false;
}

-(void)disconnected
{
    [self->connectDialog dismissWithClickedButtonIndex:0 animated:true];
    self->connectDialog = nil;
    self.tabBarController.selectedIndex = 0;
    
    //self->searchEnabled = true;
    
    //[self searchEnded];
}

-(NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

-(NSString*)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
    return @"";
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self->devices.count;
}

-(UITableViewCell*)tableView:(UITableView*)tableView1 cellForRowAtIndexPath:(NSIndexPath*)indexPath
{
    DiscoveredRC* item = [self->devices objectAtIndex:indexPath.row];
    
    static NSString* cellIdentifier = @"ItemCell";
    
    UITableViewCell* cell = [tableView1 dequeueReusableCellWithIdentifier:cellIdentifier];
    
    if (cell == nil)
    {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:cellIdentifier];
    }
    
    if (nil != item.deviceName)
    {
        cell.textLabel.text = item.deviceName;
    }
    
    if (nil != item.deviceAddress)
    {
        cell.detailTextLabel.text = item.deviceAddress;
    }
    
    cell.accessoryType = UITableViewCellAccessoryNone;
    cell.selectionStyle = UITableViewCellSelectionStyleGray;
    
    return cell;
}

-(void)tableView:(UITableView*)tableView1 didSelectRowAtIndexPath:(NSIndexPath*)indexPath
{
}

-(BOOL)tableView:(UITableView*)tableView1 canEditRowAtIndexPath:(NSIndexPath*)indexPath
{
    return YES;
}

-(void)tableView:(UITableView*)tableView1 commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath*)indexPath
{
}

-(BOOL)tabBarController:(UITabBarController*)tabBarController shouldSelectViewController:(UIViewController*)viewController
{
    bool connected = (nil != [CarsController getInstance].connectedCar);
    
    if (viewController == self)
    {
        return !connected;
    }
    else if (viewController.class == SettingsViewController.class)
    {
        return connected;
    }
    else if (viewController.class == StatusViewController.class)
    {
        return connected;
    }
    else if (viewController.class == CarViewController.class)
    {
        return connected;
    }
    
    return true;
}

@end

File: software/ios/ble-rc/ConnectViewController.swift
import Foundation
import UIKit

class ConnectViewControllerEx : UIViewController, UITableViewDelegate, UITabBarControllerDelegate, UITableViewDataSource
{
	@IBOutlet var radioCarsTableView: UITableView!;
	private var blueToothEnabled = false;
	private var refreshControl: UIRefreshControl!;
	private var connectDialog: UIAlertController!;

	@IBAction func addClicked(sender: AnyObject!)
	{
		let optionMenu = UIAlertController(title: "Add car", message: AppSettings.addMessage, preferredStyle: .Alert);

		optionMenu.addAction(UIAlertAction(title: "Get some cars", style: .Default, handler: { _ -> Void in UIApplication.sharedApplication().openURL(NSURL(string: AppSettings.shopForCarsUrl)!); }));

		optionMenu.addAction(UIAlertAction(title: "Add fake demo cars", style: .Default, handler: { _ -> Void in DevicesManager.sharedInstance.addFakes(); }));
		optionMenu.addAction(UIAlertAction(title: "Remove fake demo cars", style: .Default, handler: { _ -> Void in DevicesManager.sharedInstance.removeFakes(); }));
		optionMenu.addAction(UIAlertAction(title: "Cancel", style: .Cancel, handler: nil));

		self.presentViewController(optionMenu, animated: true, completion: nil);
	}

	func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int
	{
		if (true == self.blueToothEnabled)
		{
			return DevicesManager.sharedInstance.radioCars.count;
		}
		else
		{
			return 0;
		}
	}
	
	func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell
	{
		var cell: UITableViewCell! = tableView.dequeueReusableCellWithIdentifier("ItemCell");

		if (nil == cell)
		{
			cell = UITableViewCell(style: .Subtitle, reuseIdentifier: "ItemCell")
		}

		let carItem = DevicesManager.sharedInstance.radioCars[indexPath.row];

		cell.textLabel?.text = carItem.settings.name;

		if (carItem === DevicesManager.sharedInstance.activeCar)
		{
			cell.detailTextLabel?.text = "Already connected";
		}
		else
		{
			if (carItem.comment.characters.count > 0)
			{
				cell.detailTextLabel?.text = carItem.comment;
			}
			else
			{
				cell.detailTextLabel?.text = carItem.settings.type.description;
			}
		}

		cell.selectionStyle = .None;
		cell.accessoryType = .DisclosureIndicator;
		
		return cell;
	}

	override func viewDidLoad()
	{
		super.viewDidLoad();

		self.tabBarController?.delegate = self;

		DevicesManager.sharedInstance.searchEnded = self.searchEnded;
		DevicesManager.sharedInstance.bluetoothTurnedOff = self.bluetoothTurnedOff;
		DevicesManager.sharedInstance.bluetoothTurnedOn = self.bluetoothTurnedOn;
		DevicesManager.sharedInstance.deviceFound = self.reloadTable;
		DevicesManager.sharedInstance.connected = self.connected;
		DevicesManager.sharedInstance.disconnected = self.disconnected;

		self.refreshControl = UIRefreshControl();
		self.refreshControl.backgroundColor = UIColor(white: 230.0 / 255.0, alpha: 1.0);
		self.refreshControl.addTarget(self, action: "searchClicked", forControlEvents: .ValueChanged);

		let tableViewController = UITableViewController();
		tableViewController.tableView = self.radioCarsTableView;
		tableViewController.refreshControl = self.refreshControl;
	}

	func searchClicked()
	{
		DevicesManager.sharedInstance.startDeviceSearch();
	}

	private func reloadTable()
	{
		dispatch_async(dispatch_get_main_queue(),
		{
			self.radioCarsTableView?.reloadData()
		});
	}

	private func bluetoothTurnedOn()
	{
		self.blueToothEnabled = true;
		self.refreshControl.enabled = true;

		self.radioCarsTableView.backgroundView = nil;
		self.radioCarsTableView.separatorStyle = .SingleLine;
		self.refreshControl.alpha = 1;

		searchClicked();
	}

	private func bluetoothTurnedOff()
	{
		self.blueToothEnabled = false;
		self.refreshControl.enabled = false;

		DevicesManager.sharedInstance.disconnectActive();

		self.tabBarController!.selectedIndex = 0;

		self.radioCarsTableView.backgroundView = UILabel(frame: self.radioCarsTableView.bounds);
		(self.radioCarsTableView.backgroundView as! UILabel).text = "This app requires Bluetooth\nPlease enable Bluetooth.";
		(self.radioCarsTableView.backgroundView as! UILabel).numberOfLines = 0;
		(self.radioCarsTableView.backgroundView as! UILabel).textAlignment = .Center;
		self.radioCarsTableView.separatorStyle = .None;
		self.refreshControl.alpha = 0;

		reloadTable();
	}

	private func searchEnded()
	{
		if ((DevicesManager.sharedInstance.radioCars.count <= 0) && (true == self.blueToothEnabled))
		{
			let fakeCar = RadioCar(name: "Virtual Demo Car", comment: "No cars found. Please pull down to refresh.");

			DevicesManager.sharedInstance.radioCars.append(fakeCar);

			reloadTable();
		}

		if (nil != self.refreshControl)
		{
			self.refreshControl.endRefreshing();
		}
	}

	private func connected()
	{
		if (nil != self.connectDialog)
		{
			self.connectDialog.dismissViewControllerAnimated(true, completion: {});
			self.connectDialog = nil;
		}

		self.tabBarController!.selectedIndex = 1;
	}

	private func disconnected()
	{
		if (nil != self.connectDialog)
		{
			self.connectDialog.dismissViewControllerAnimated(true, completion: {});
			self.connectDialog = nil;
		}

		self.tabBarController!.selectedIndex = 0;
	}

	func tabBarController(tabBarController: UITabBarController, shouldSelectViewController viewController: UIViewController) -> Bool
	{
		let connected = (nil != DevicesManager.sharedInstance.activeCar);

		if (viewController.tabBarItem.tag == 22)
		{
			return true;
		}
		else if (viewController.tabBarItem.tag == 11)
		{
			return connected;
		}

		return true;
	}

	func tabBarController(tabBarController: UITabBarController, didSelectViewController viewController: UIViewController)
	{
		if (viewController.tabBarItem.tag == 22)
		{
			searchClicked();
		}
	}

	private func connect(selectedIndexPath: NSIndexPath!)
	{
		if (nil != selectedIndexPath)
		{
			if ((selectedIndexPath.row >= 0) && (selectedIndexPath.row < DevicesManager.sharedInstance.radioCars.count))
			{
				let carItem = DevicesManager.sharedInstance.radioCars[selectedIndexPath.row];

				if (carItem === DevicesManager.sharedInstance.activeCar)
				{
					self.tabBarController!.selectedIndex = 1;
					return;
				}

				if (nil == self.connectDialog)
				{
					self.connectDialog = UIAlertController(title: nil, message: "Connecting to " + carItem.settings.name + "...\n\n\n", preferredStyle: .Alert)

					let indicator = UIActivityIndicatorView(frame: self.connectDialog.view.bounds)
					indicator.autoresizingMask = [ UIViewAutoresizing.FlexibleWidth, UIViewAutoresizing.FlexibleHeight ];
					indicator.tintColor = AppSettings.applicationTint;
					self.connectDialog.view.addSubview(indicator);
					indicator.userInteractionEnabled = false;
					indicator.startAnimating();

					self.connectDialog.addAction(UIAlertAction(title: "Cancel", style: .Cancel, handler: { _ in DevicesManager.sharedInstance.cancelConnection(carItem); }));
					self.presentViewController(self.connectDialog, animated: true, completion: nil)
				}

				DevicesManager.sharedInstance.connect(carItem);
			}
		}
	}

	func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath)
	{
		connect(indexPath);
	}
}
